arangod/Aql/ExecutionPlan.cpp

```diff
@@ -2131,47 +2137,39 @@ ExecutionNode* ExecutionPlan::fromNode(AstNode const* node) {
   return en;
 }
 
-/// @brief find nodes of a certain type
+template <bool unique>
+/// @brief find nodes of certain types
 void ExecutionPlan::findNodesOfType(::arangodb::containers::SmallVector<ExecutionNode*>& result,
-                                    ExecutionNode::NodeType type, bool enterSubqueries) {
-  // consult our nodes-of-type counters array
-  if (contains(type)) {
-    // node type present in plan
-    NodeFinder<ExecutionNode::NodeType> finder(type, result, enterSubqueries);
+                                    std::initializer_list<ExecutionNode::NodeType> const& types,
+                                    bool enterSubqueries) {
+  // check if any of the node types is actually present in the plan
+  Contains contains(*this);
+  bool haveNodes = std::any_of(types.begin(), types.end(), contains);
```

Comment:
I have to say I find this less readable than a simple
```
bool haveNodes = std::any_of(types.begin(), types.end(), [this](ExecutionNode::NodeType type) {
  return contains(type);
});
```
But with the separate `Contains` I have to look elsewhere what it actually does.

Reply:
`this` should just cost 8 bytes, and with the g++ implementation lambdas should use an SSO with a 16 bytes buffer. So you could even capture two pointer values without getting into the dynamic memory allocation case.
I am not 100% sure about MSVC, but I guess it has a reasonable optimization for this common case too.

Reply:
Okay, then I'll go ahead and convert to the lambda. I agree that it is definitely more readable.

Reply:
Btw., what I wrote is true for `std::function`. It seems for lambdas the compiler is even smarter. Need to compile a bit more data...

Reply:
`any_of` & co take the functor as template parameter by value, so the anonymous struct of the lambda is copied/moved into the function, so no dynamic allocation is necessary.

Reply:
FWIW, with respect to compiler optimization, the lambda is actually preferable. In `std::any_of(types.begin(), types.end(), contains)` the third parameter `contains` decays to a function pointer, but a function pointer is much harder to inline. With the lambda, the full type information of the (anonymous)  functor is available to the compiler, and that type information carries with it the code needed to be inlined.